File: backend/src/main/java/com/manarah/common/SchoolYears.java
package com.manarah.common;

import java.util.Map;

/**
 * School years are free text: a teacher types "الصف الثاني الثانوي" on a course, a student picked "تانية ثانوي",
 * an older account says "2 ثانوي". {@link #key} reduces any of those to one comparable key ("secondary-2"), so a
 * student sees exactly the courses of their own year however either side spelled it. Text that names no stage and
 * number (a course "لكل السنين") falls back to its normalized form, which only matches the same wording.
 */
public final class SchoolYears {
    private SchoolYears() {}

    private static final Map<String, Integer> ORDINALS = Map.ofEntries(
            Map.entry("اول", 1), Map.entry("اولي", 1), Map.entry("اولا", 1),
            Map.entry("ثاني", 2), Map.entry("ثانيه", 2), Map.entry("تاني", 2), Map.entry("تانيه", 2),
            Map.entry("ثالث", 3), Map.entry("ثالثه", 3), Map.entry("تالت", 3), Map.entry("تالته", 3),
            Map.entry("رابع", 4), Map.entry("رابعه", 4),
            Map.entry("خامس", 5), Map.entry("خامسه", 5),
            Map.entry("سادس", 6), Map.entry("سادسه", 6));

    /** Comparable key for a school year, or "" for blank input. */
    public static String key(String raw) {
        String s = normalize(raw);
        if (s.isEmpty()) return "";
        String stage = s.contains("ثانوي") ? "secondary" : s.contains("اعدادي") ? "preparatory"
                : s.contains("ابتدائي") ? "primary" : s.contains("روضه") || s.contains("kg") ? "kg" : null;
        Integer number = null;
        for (String token : s.split(" ")) {
            // "3ث" / "2ع": the usual shorthand for third secondary / second preparatory.
            if (token.matches("[1-6][ثعب]")) {
                number = token.charAt(0) - '0';
                if (stage == null) stage = switch (token.charAt(1)) { case 'ث' -> "secondary"; case 'ع' -> "preparatory"; default -> "primary"; };
                break;
            }
            if (token.matches("[1-6]")) { number = Integer.parseInt(token); break; }
            String word = token.startsWith("ال") && token.length() > 3 ? token.substring(2) : token;
            if (ORDINALS.containsKey(word)) { number = ORDINALS.get(word); break; }
        }
        // "الثانوية العامة" is how everyone says the final secondary year.
        if ("secondary".equals(stage) && number == null && s.contains("عامه")) number = 3;
        if (stage != null && number != null) return stage + "-" + number;
        return s.replaceAll("(^| )الصف( |$)", " ").trim();
    }

    /** "للصف الثاني الثانوي" / "لـتانية ثانوي": "for <year>" the way it's written in Arabic. */
    public static String forYear(String year) {
        return year.startsWith("ال") ? "لل" + year.substring(2) : "لـ" + year;
    }

    /** Whether two years are the same year. Blank never matches anything. */
    public static boolean same(String a, String b) {
        String ka = key(a);
        return !ka.isEmpty() && ka.equals(key(b));
    }

    /** Folds the spelling variations that don't change meaning: hamza forms, ى/ي, ة/ه, tatweel, diacritics, digits. */
    static String normalize(String raw) {
        if (raw == null) return "";
        StringBuilder out = new StringBuilder(raw.length());
        for (char c : raw.trim().toLowerCase().toCharArray()) {
            if (c >= 'ً' && c <= 'ْ' || c == 'ـ') continue;
            if (c >= '٠' && c <= '٩') c = (char) ('0' + (c - '٠'));
            else if (c >= '۰' && c <= '۹') c = (char) ('0' + (c - '۰'));
            else if (c == 'أ' || c == 'إ' || c == 'آ' || c == 'ٱ') c = 'ا';
            else if (c == 'ى') c = 'ي';
            else if (c == 'ة') c = 'ه';
            else if (c == '-' || c == '_' || c == '/' || c == '،' || c == ',') c = ' ';
            out.append(c);
        }
        // "2ثانوي" → "2 ثانوي", so the number and the stage read as separate words.
        return out.toString().replaceAll("([0-9])([^0-9 ][^ ]{2,})", "$1 $2").replaceAll("\\s+", " ").trim();
    }
}

File: frontend/src/lib/schoolYears.js
// Mirror of the backend's SchoolYears: school years are free text ("الصف الثاني الثانوي", "تانية ثانوي", "2ث"),
// so years are compared by a key, never by exact spelling. Keep the two in step.

const ORDINALS = {
  اول: 1, اولي: 1, اولا: 1,
  ثاني: 2, ثانيه: 2, تاني: 2, تانيه: 2,
  ثالث: 3, ثالثه: 3, تالت: 3, تالته: 3,
  رابع: 4, رابعه: 4, خامس: 5, خامسه: 5, سادس: 6, سادسه: 6,
}

function normalize(raw) {
  if (!raw) return ''
  let out = ''
  for (let c of String(raw).trim().toLowerCase()) {
    const code = c.charCodeAt(0)
    if ((code >= 0x064b && code <= 0x0652) || code === 0x0640) continue
    if (code >= 0x0660 && code <= 0x0669) c = String(code - 0x0660)
    else if (code >= 0x06f0 && code <= 0x06f9) c = String(code - 0x06f0)
    else if ('أإآٱ'.includes(c)) c = 'ا'
    else if (c === 'ى') c = 'ي'
    else if (c === 'ة') c = 'ه'
    else if ('-_/،,'.includes(c)) c = ' '
    out += c
  }
  return out.replace(/([0-9])([^0-9 ][^ ]{2,})/g, '$1 $2').replace(/\s+/g, ' ').trim()
}

/** Comparable key for a school year ("secondary-2"), or '' for blank input. */
export function yearKey(raw) {
  const s = normalize(raw)
  if (!s) return ''
  let stage = s.includes('ثانوي') ? 'secondary' : s.includes('اعدادي') ? 'preparatory'
    : s.includes('ابتدائي') ? 'primary' : s.includes('روضه') || s.includes('kg') ? 'kg' : null
  let number = null
  for (const token of s.split(' ')) {
    if (/^[1-6][ثعب]$/.test(token)) {
      number = Number(token[0])
      if (!stage) stage = token[1] === 'ث' ? 'secondary' : token[1] === 'ع' ? 'preparatory' : 'primary'
      break
    }
    if (/^[1-6]$/.test(token)) { number = Number(token); break }
    const word = token.startsWith('ال') && token.length > 3 ? token.slice(2) : token
    if (ORDINALS[word]) { number = ORDINALS[word]; break }
  }
  if (stage === 'secondary' && number == null && s.includes('عامه')) number = 3
  if (stage && number) return `${stage}-${number}`
  return s.replace(/(^| )الصف( |$)/g, ' ').trim()
}

/** Whether two years are the same year. Blank never matches. */
export const sameYear = (a, b) => { const k = yearKey(a); return !!k && k === yearKey(b) }

/** "للصف الثاني الثانوي" / "لـتانية ثانوي": "for <year>" the way it's written in Arabic. */
export const forYear = (y) => (y.startsWith('ال') ? `لل${y.slice(2)}` : `لـ${y}`)

/** One entry per distinct year, keeping the first spelling seen. */
export function distinctYears(list) {
  const seen = new Map()
  for (const y of list) if (y && !seen.has(yearKey(y))) seen.set(yearKey(y), String(y).trim())
  return [...seen.values()]
}
